src/high-level/abstract-tensor.lisp

```diff
@@ -0,0 +1,269 @@
+;;;; abstract-tensor.lisp
+;;;;
+;;;; Author: Cole Scott
+;;;;
+;;;; Collaborators: Robert Smith
+
+(in-package #:magicl)
+
+(defstruct (abstract-tensor))
+
+;;; abstract-tensor protocol
+;;; Must be implemented by subclasses
+
+(defgeneric shape (tensor)
+  (:documentation "The shape (dimensions) of the tensor. eg. '(2 3) for a 2x3 tensor"))
+
+(defgeneric tref (tensor &rest pos)
+  (:documentation "Get a reference to element at position"))
+
+(defgeneric (setf tref) (new-value tensor &rest pos)
+  (:documentation "Set the value of element at position"))
+
+(defgeneric copy-tensor (tensor &rest args)
+  (:documentation "Create a new tensor with the same properties as the given tensor, not initializing the storage"))
+
+(defgeneric deep-copy-tensor (tensor &rest args)
+  (:documentation "Create a new tensor with the same properties as the given tensor, copying storage"))
+
+;;; abstract-tensor generic methods
+;;; Can be optimized by subclasses
```

Comment:
I would elaborate. "These have default generic implementations but can be optimized by subclasses."